tests/h/auth/policy_test.py

```diff
@@ -113,43 +113,232 @@ def test_forget_uses_api_policy_for_api_paths(self, api_request):
-        check.return_value = []
-        policy = AuthClientPolicy(check=check)
+    def test_it_instantiates_a_BasicAuthAuthenticationPolicy(self, BasicAuthAuthenticationPolicy):
+        policy = AuthClientPolicy()
 
-        policy.effective_principals(pyramid_request)
+        assert policy.basic_auth_policy == BasicAuthAuthenticationPolicy.return_value
+        BasicAuthAuthenticationPolicy.assert_called_once_with(check=AuthClientPolicy.check)
 
-        check.assert_called_once_with("foo", "bar", pyramid_request)
+    def test_forwarded_userid_defaults_to_checking_XForwardedUser_header(self, pyramid_request):
+        pyramid_request.headers['X-Forwarded-User'] = 'filbert'
 
-    def test_it_returns_None_for_authenticated_userid(self, pyramid_request, check):
-        check.return_value = []
+        assert AuthClientPolicy.forwarded_userid(pyramid_request) == 'filbert'
+
+    def test_forwarded_userid_returns_None_if_header_not_set(self, pyramid_request):
+        assert AuthClientPolicy.forwarded_userid(pyramid_request) is None
+
+    def test_unauthenticated_userid_returns_forwarded_user_if_present(self, auth_policy, pyramid_request):
+        pyramid_request.headers['X-Forwarded-User'] = 'filbert'
+
+        userid = auth_policy.unauthenticated_userid(pyramid_request)
+
+        assert userid == 'filbert'
+
+    def test_unauthenticated_userid_returns_clientid_if_no_forwarded_user(self, auth_policy, pyramid_request):
+        userid = auth_policy.unauthenticated_userid(pyramid_request)
+
+        assert userid == 'foo'
+
+    def test_unauthenticated_userid_proxies_to_basic_auth_if_no_forwarded_user(self,
+                                                                               auth_policy,
+                                                                               pyramid_request,
+                                                                               BasicAuthAuthenticationPolicy):
+        auth_policy.basic_auth_policy = BasicAuthAuthenticationPolicy(check=AuthClientPolicy.check)
+
+        auth_policy.unauthenticated_userid(pyramid_request)
+
+        auth_policy.basic_auth_policy.unauthenticated_userid.assert_called_once_with(pyramid_request)
+
+    def test_unauthenticated_userid_doesnt_proxy_to_basic_auth_if_forwarded_user(self,
+                                                                                auth_policy,
+                                                                                pyramid_request,
+                                                                                BasicAuthAuthenticationPolicy):
+        pyramid_request.headers['X-Forwarded-User'] = 'dingbat'
+        auth_policy.basic_auth_policy = BasicAuthAuthenticationPolicy(check=AuthClientPolicy.check)
+
+        auth_policy.unauthenticated_userid(pyramid_request)
+
+        auth_policy.basic_auth_policy.unauthenticated_userid.assert_not_called()
```

Comment:
This land mine again, `assert_not_called()` doesn't exist in the Python 2 / third-party mock library! And `auth_policy.basic_auth_policy` is an unconstrained mock. The `BasicAuthAuthenticationPolicy` mock is created using autospeccing (via our `patch` fixture, which uses autospec) but unfortunately [autospec doesn't spec the return values of mock objects and their methods](https://hyp.is/cfYO-JdGEeiL-P8E8GQW1g/www.seanh.cc/posts/the-problem-with-mocks) (it can't, because mock would have to actually call our methods in order to see what shape of object they return). So while the `BasicAuthAuthenticationPolicy` is autospec'd, object instances of that class (i.e. the mock class's return values) are unconstrained mocks.

You might think autospec could maybe detect the thing being mocked is a class and autospec the class's return value / the object instances as well, but unfortunately there are reasons why that would cause problems.

The tests can fix this like this by doing mock's work for it:

```python
@pytest.fixture
def BasicAuthAuthenticationPolicy(self, patch):
    BasicAuthAuthenticationPolicy = patch("h.auth.policy.BasicAuthAuthenticationPolicy")
    BasicAuthAuthenticationPolicy.return_value = mock.create_autospec(BasicAuthAuthenticationPolicy, instance=True, spec_set=True)
```